crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1340,21 +1365,77 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Resolves the length value of a type.
+    ///
+    /// This is used to determine the value that would be returned
+    /// when `len(x)` is called on an object `x`.
+    fn len(&self, db: &'db dyn Db) -> Option<Type<'db>> {
+        match self {
+            Type::BytesLiteral(bytes) => return bytes.len(db).try_into().ok(),
+            Type::StringLiteral(string) => {
+                return string.as_str(db).chars().count().try_into().ok()
+            }
+
+            Type::Tuple(tuple) => return tuple.elements(db).len().try_into().ok(),
+
+            ty if !matches!(ty, Type::Instance(..)) => return None,
+
+            _ => {}
+        }
+
+        let Type::Instance(instance) = self else {
+            return None;
+        };
+        let Symbol::Type(dunder_len, _) = instance.class.class_member(db, "__len__") else {
+            return None;
+        };
+        let Type::FunctionLiteral(function) = dunder_len else {
+            return None;
+        };
```

Comment:
It doesn't actually need to be a function; any callable object will do:

```pycon
>>> class Foo:
...     def __call__(self):
...         return 42
...         
>>> class Bar:
...     __len__ = Foo()
...     
>>> len(Bar())
42
```

We have a `Type::call()` method that abstracts away the details of trying to call a possibly-not-callable type; you should be able to use that here, which will mean you'll support arbitrary callable objects here rather than just functions